### PR TITLE
Cancel Order

### PR DESCRIPTION
**Pull request by husobee**
### Summary

adding ability to cancel an order, if cancelled, and stripe subscription, cancel subscription

### Type of change ( select one )

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: payment/controllers.go [hunk 1]
```diff
@@ -56,6 +56,7 @@ func Router(service *Service) chi.Router {
 
 	r.Method("OPTIONS", "/{orderID}", middleware.InstrumentHandler("GetOrderOptions", corsMiddleware([]string{"GET"})(nil)))
 	r.Method("GET", "/{orderID}", middleware.InstrumentHandler("GetOrder", corsMiddleware([]string{"GET"})(GetOrder(service))))
+	r.Method("DELETE", "/{orderID}", middleware.InstrumentHandler("CancelOrder", corsMiddleware([]string{"DELETE"})(CancelOrder(service))))
```
Comment (evq) [inline payment/controllers.go:59]: I don't think we should have cancel order without authentication

**Review by evq — CHANGES_REQUESTED** [review]

File: payment/controllers.go (hunk 1 above)
Comment (husobee) [reply payment/controllers.go:59]: Simple token should do right?  This is initiated from subscriptions?

File: payment/datastore.go [hunk 2]
```diff
@@ -350,9 +353,73 @@ func (pg *Postgres) GetTransaction(externalTransactionID string) (*Transaction,
...
+		}
+	}
+	return ok, "", err
+}
+
 // UpdateOrder updates the orders status.
 // 	Status should either be one of pending, paid, fulfilled, or canceled.
 func (pg *Postgres) UpdateOrder(orderID uuid.UUID, status string) error {
+	// if the order is cancelled we might need to inform other systems
+	if status == "canceled" {
```
Comment (evq) [inline payment/datastore.go:409]: it seems strange to have this logic live in the db, can we instead have a non-handler cancel method on the service which both updates the db and performs this?

Comment (yachtcaptain23) [reply payment/datastore.go:409]: Can we make `canceled` an enum, as `canceled` and `cancelled` are both valid English words?

Comment (evq) [reply payment/datastore.go:409]: what's an enum, this is golang

File: cmd/macaroon/brave-talk/premium_prod.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+      - description: "Premium access to Brave Talk"
+      - credential_type: "single-use"
+      - allowed_payment_methods: "stripe"
+      - metadata: '
+{
+  "stripe_product_id": "prod_Jw4zQxdHkpxSOe",
+  "stripe_item_id": "price_1JICpEBSm1mtrN9nwCKvpYQ4",
+  "stripe_success_uri": "https://account.brave.com/account/?intent=provision",
+  "stripe_cancel_uri": "https://account.brave.com/plans/?intent=checkout"
+}'
```
Comment (yachtcaptain23) [inline cmd/macaroon/brave-talk/premium_prod.yaml:18]:
```suggestion
tokens:
    id: "brave-talk-premium-prod sku token v1"
    version: 1
    location: "talk.brave.com"
    first_party_caveats:
      sku: "brave-talk-premium-prod"
      price: 7.00
      currency: "USD"
      description: "Premium access to Brave Talk"
      credential_type: "single-use"
      allowed_payment_methods: "stripe"
      metadata: >-
                        {
                           "stripe_product_id": "prod_Jw4zQxdHkpxSOe",
                           "stripe_item_id": "price_1JICpEBSm1mtrN9nwCKvpYQ4",
                           "stripe_success_uri": "https://account.brave.com/account/?intent=provision",
                           "stripe_cancel_uri": "https://account.brave.com/plans/?intent=checkout"
                        }
```

(1) I think the dashes were added in the wrong place (they usually instantiate default values in an array.)
(2) I referenced https://stackoverflow.com/questions/3790454/how-do-i-break-a-string-in-yaml-over-multiple-lines to make a better looking multi-line JSON

File: payment/datastore.go (hunk 2 above)
Comment (husobee) [reply payment/datastore.go]: resolved both comments: https://github.com/brave-intl/bat-go/pull/889/commits/df62ffdda5dfffd1eea14232563c78228af9d6c9

File: payment/controllers.go (hunk 1 above)
Comment (husobee) [reply payment/controllers.go]: added authentication: https://github.com/brave-intl/bat-go/pull/889/commits/df62ffdda5dfffd1eea14232563c78228af9d6c9

**Review by evq — APPROVED** [review]

**Pull request merged**